### PR TITLE
role/chore: feat to run fluentd as another group

### PR DESCRIPTION
**Pull request by StessyMougeot**

### CODE REVIEW

**Comment by ghost** [comment]
Un petit update dans le README.md serait le bienvenu, ainsi qu'un fichier changelog et un tag, comme pour les autres rôles ;)

File: tasks/main.yml [hunk 1]
```diff
@@ -10,6 +10,14 @@
     name: td-agent
     state: "{{ fluentd_package_state }}"
 
+- name: Change group to adm
```
Comment (sterchelen) [inline tasks/main.yml]:
```suggestion
- name: Override fluentd default group to {{ fluentd_group }}
```

File: tasks/main.yml [hunk 2]
```diff
@@ -10,6 +10,14 @@
...
+  lineinfile:
+    path: /etc/systemd/system/multi-user.target.wants/td-agent.service
+    regexp: '^Group='
+    line: 'Group={ fluentd_group }'
```
Comment (sterchelen) [inline tasks/main.yml]: If you look closely to this unit file
```
[Unit]
Description=td-agent: Fluentd based data collector for Treasure Data
Documentation=https://docs.treasuredata.com/articles/td-agent
After=network-online.target
Wants=network-online.target

[Service]
User=td-agent
Group=adm
LimitNOFILE=65536
Environment=LD_PRELOAD=/opt/td-agent/lib/libjemalloc.so
Environment=GEM_HOME=/opt/td-agent/lib/ruby/gems/2.7.0/
Environment=GEM_PATH=/opt/td-agent/lib/ruby/gems/2.7.0/
Environment=FLUENT_CONF=/etc/td-agent/td-agent.conf
Environment=FLUENT_PLUGIN=/etc/td-agent/plugin
Environment=FLUENT_SOCKET=/var/run/td-agent/td-agent.sock
Environment=TD_AGENT_LOG_FILE=/var/log/td-agent/td-agent.log
Environment=TD_AGENT_OPTIONS=
EnvironmentFile=-/etc/default/td-agent
PIDFile=/var/run/td-agent/td-agent.pid
RuntimeDirectory=td-agent
Type=forking
# XXX: Fix fluentd executables path
ExecStart=/opt/td-agent/bin/fluentd --log $TD_AGENT_LOG_FILE --daemon /var/run/td-agent/td-agent.pid $TD_AGENT_OPTIONS
ExecStop=/bin/kill -TERM ${MAINPID}
ExecReload=/bin/kill -HUP ${MAINPID}
Restart=always
TimeoutStopSec=120

[Install]
WantedBy=multi-user.target
```

you have this mention `EnvironmentFile=-/etc/default/td-agent`
Let's add the group into the `/etc/default/td-agent` file. Would be much more maintainable :)

**Review by sterchelen — CHANGES_REQUESTED** [review]

File: tasks/main.yml (hunk 2 above)
Comment (StessyMougeot) [reply tasks/main.yml]: Not possible, it's an escalation privileges

Comment (sterchelen) [reply tasks/main.yml]: Seen with you privately. Let's add `td-agent` user to the group `adm`. This way, we don't have to modify the service file :smiley:

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Changelog
+=========
+
```
Comment (sterchelen) [inline CHANGELOG.md:3]: Could you add `0.1.1` with the actual changes?

File: README.md [hunk 4]
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
 
     fluentd_package_state: present
 
+Extra groups to add log access, secret:
```
Comment (sterchelen) [inline README.md]: What you mean by `, secret`?

File: README.md [hunk 5]
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
...
 
+    fluentd_adm_group: false
     fluentd_service_name: td-agent
     fluentd_service_state: started
     fluentd_service_enabled: true
 
+Vault integration to use alongside `aggregator` tag to load secret in environment variable:
+    fluentd_vault_config_path:
+    fluentd_secret_path:
+    fluent_secret_name:
```
Comment (sterchelen) [inline README.md]:
```suggestion
    fluentd_secret_name:
```

File: tasks/main.yml [hunk 6]
```diff
@@ -10,6 +10,15 @@
     name: td-agent
     state: "{{ fluentd_package_state }}"
 
+- name: "Add user to {{ fluentd_groups | join(',') }} group"
+  user:
+    name: td-agent
+    group: td-agent
```
Comment (sterchelen) [inline tasks/main.yml:16]: Not sure if it is mandatory?

File: tasks/main.yml [hunk 7]
```diff
@@ -28,8 +37,54 @@
     user_install: false
   with_items: "{{ fluentd_plugins }}"
 
+- name: retrieve initial ExecStart
+  shell: 'grep -i "ExecStart=.*/.*" /lib/systemd/system/multi-user.target.wants/td-agent.service | sed -r "s/^([^\/])*(\/.*)/\2/"'
```
Comment (sterchelen) [inline tasks/main.yml]: Is it possible to use `/lib/systemd/system/td-agent.service`? We should use the original service file

File: tasks/main.yml [hunk 8]
```diff
@@ -28,8 +37,54 @@
...
+  - never
+  - aggregator
+
+- name: Ensure override file is absent
+  file:
+    path: /etc/systemd/system/td-agent.service.d/override.conf
+    state: absent
+  tags:
+  - never
+  - aggregator
```
Comment (sterchelen) [inline tasks/main.yml]: Is it mandatory to ensure if the file is absent?

File: tasks/main.yml [hunk 9]
```diff
@@ -28,8 +37,54 @@
...
+
+- name: Add vault secret to ExecStart
+  lineinfile:
+    path: /etc/systemd/system/td-agent.service.d/override.conf
+    state: present
+    insertafter: '^ExecStart=$'
+    line: 'ExecStart=vault --config-file {{ fluentd_vault_config_path }} env -p {{ fluentd_secret_path }}{{ fluent_secret_name }} -- {{ initial_cmd.stdout }}'
```
Comment (sterchelen) [inline tasks/main.yml]: To improve readability:
```suggestion
    line: > 
    'ExecStart=vault --config-file {{ fluentd_vault_config_path }}
    env -p {{ fluentd_secret_path }}{{ fluent_secret_name }} -- {{ initial_cmd.stdout }}'
```

File: README.md (hunk 4 above)
Comment (StessyMougeot) [reply README.md]: For exemple, to read datadog config file for vault, user must be in `vault-datadog` group

File: tasks/main.yml (hunk 8 above)
Comment (StessyMougeot) [reply tasks/main.yml]: Is there a better way to clear the file with ansible?

Comment (sterchelen) [reply tasks/main.yml]: Yes, using a template file :)

Comment (sterchelen) [reply tasks/main.yml]: Then you can remove merge `reset ExecStart` task with `Add vault secret`

File: tasks/main.yml (hunk 6 above)
Comment (sterchelen) [reply tasks/main.yml:16]: I mean specifying `group: td-agent`

File: README.md [hunk 10]
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
 
     fluentd_package_state: present
 
+Extra groups to add log access, secret:
+    fluentd_groups:
```
Comment (sterchelen) [inline README.md]: Your naming `extra groups` speaks by itself:
```suggestion
    fluentd_extra_groups:
```

File: README.md (hunk 4 above)
Comment (sterchelen) [reply README.md]: Then I would rephrase like this:
```suggestion
Extra groups to give specific accesses to fluentd user, like logs or secret:
```

File: README.md [hunk 11]
```diff
@@ -14,12 +14,22 @@ Available variables are listed below, along with default values (see `defaults/m
...
+    - adm
+
 The `td-agent` Fluentd package state; set to `latest` to upgrade or change versions.
 
+    fluentd_adm_group: false
```
Comment (sterchelen) [inline README.md]: I am not seeing where you are using it :eyes:

File: tasks/main.yml (hunk 6 above)
Comment (StessyMougeot) [reply tasks/main.yml:16]: I don't think it's mandatory, but for informations about primary group

File: tasks/main.yml (hunk 7 above)
Comment (StessyMougeot) [reply tasks/main.yml]: Sure, copy/paste mistake

File: README.md (hunk 11 above)
Comment (StessyMougeot) [reply README.md]: Not anymore, incorporated in `fluentd_extra_groups`

File: README.md (hunk 10 above)
Comment (StessyMougeot) [reply README.md]: I've hesitated between both in fact

File: tasks/main.yml (hunk 8 above)
Comment (StessyMougeot) [reply tasks/main.yml]: Much better, thanks

File: tasks/main.yml [hunk 12]
```diff
@@ -28,8 +37,33 @@
...
+    path: /etc/systemd/system/td-agent.service.d
+    state: directory
+    mode: '0755'
+  tags:
+  - never
+  - aggregator
+
+- name: Add unit override.conf file
+  template:
+    src: templates/override.conf.j2
```
Comment (sterchelen) [inline tasks/main.yml:58]: You've forgotten to add the template file :smiley:

**Review by sterchelen — APPROVED** [review]
